src/pupil_labs/invisible_lsl_relay/linear_time_model.py

```diff
@@ -0,0 +1,149 @@
+        if (
+            event_stream['info']['desc'][0]['acquisition'][0]['world_camera_serial'][0]
+            == world_camera_serial_num
+        ):
+            return xdf_event_to_df(event_stream)
+    raise ValueError(
+        f'None of streams in {path_to_xdf_recording} contained '
+        f'the serial number {world_camera_serial_num}!'
+    )
+
+
+def xdf_event_to_df(
+    event_stream, column_name=event_column_name, column_timestamp=event_column_timestamp
+):
+    lsl_event_data = pd.DataFrame(columns=[column_name, column_timestamp])
+    lsl_event_data[column_name] = [name[0] for name in event_stream['time_series']]
+    lsl_event_data[column_timestamp] = event_stream['time_stamps']
+
+    return lsl_event_data
+
+
+def load_cloud_data(cloud_path, column_timestamp=event_column_timestamp):
+    cloud_event_data = load_df_from_dir(cloud_path, 'events.csv')
+    # transform cloud timestamps to seconds
+    cloud_event_data[column_timestamp] = cloud_event_data['timestamp [ns]'] * 1e-9
+
+    return cloud_event_data
+
+
+def filter_common_events(cloud_data, xdf_data, column_name=event_column_name):
+
+    name_intersection = np.intersect1d(cloud_data[column_name], xdf_data[column_name])
+
+    # filter timestamps by the event intersection
+    filtered_cloud_data = cloud_data[cloud_data[column_name].isin(name_intersection)]
+    filtered_xdf_data = xdf_data[xdf_data[column_name].isin(name_intersection)]
+
+    return filtered_cloud_data, filtered_xdf_data
+
+
+def fit_linear_model(
+    filtered_cloud_data,
+    filtered_xdf_data,
+    cloud_path,
+    file_name='gaze.csv',
+    column_timestamp=event_column_timestamp,
+):
+
+    cloud_to_lsl_mapper = linear_time_mapper(
+        filtered_cloud_data[[column_timestamp]], filtered_xdf_data[column_timestamp]
+    )
+    lsl_to_cloud_mapper = linear_time_mapper(
+        filtered_xdf_data[[column_timestamp]], filtered_cloud_data[column_timestamp]
+    )
+    # load the file with the gaze data
+    cloud_gaze_data = load_df_from_dir(cloud_path, file_name)
+    # map from nanoseconds to seconds
+    cloud_gaze_data[column_timestamp] = cloud_gaze_data['timestamp [ns]'] * 1e-9
+
+    # map cloud time to lsl time in seconds
+    cloud_gaze_data['lsl_time [s]'] = cloud_to_lsl_mapper.predict(
+        cloud_gaze_data[[column_timestamp]]
+    )
+
+    return cloud_gaze_data, cloud_to_lsl_mapper, lsl_to_cloud_mapper
+
+
+def linear_time_mapper(x, y):
+    mapper = linear_model.LinearRegression()
+    mapper.fit(x, y)
+
+    return mapper
+
+
+def load_df_from_dir(dir_name, file_name, expected_n=1):
+    if expected_n > 1:
+        all_dfs = []
+
+    files = list(pathlib.Path(dir_name).rglob(file_name))
+    if len(files) != expected_n:
+        raise ValueError(
+            f'There was not the correct number of event files in the '
+            f"directory {dir_name} and it's sub directories."
+            f'expected {expected_n}, but got {len(file_name)}.'
+        )
```

Comment:
I like the "search for all available files based on a pattern" option! But it might be hard to know how large `expected_n` is meant to be beforehand. Either have the user provide a list of expected files, or let them use a pattern that can match any number of files (a warning at num_matched==0 might be sensible).

Reply:
The current implementation of this works with the assumption of one info.json -> one enrichment download, so the expected n is 1. 

However, this will break if there are multiple enrichments associated with one info.json file. To fix this, further changes are required to the relay, so this is to be implemented later.